cypress/GenerateCypressTestSpecs.js

```diff
@@ -0,0 +1,147 @@
+
+var fs = require('fs')
+
+const pathToTestFiles = './cypress/tests'
+const pathToTestSpecFiles = './cypress/integration/'
+var fullTestList = []
+
+main();
+WatchAndRun('./cypress/tests', main)
+
+// ----------------------------------------------------------------------
+// Watch the a folder for changes and when one is found run a function.
+// ----------------------------------------------------------------------
+
+function WatchAndRun(pathToWatch, functionToRun) {
+  var timeoutHandle;
+  var timeOfLastChange;
+  
+  fs.watch(pathToWatch, (eventType, filename) => {
+    var now = new Date();
+    if (timeOfLastChange && now - timeOfLastChange < 1000) { clearTimeout(timeoutHandle); }
+    timeOfLastChange = now;
+    timeoutHandle = setTimeout(functionToRun, 1000)
+  });
+}
+
+// ----------------------------------------------------------------------
+
+function main() {
+  try {
+    // Get the test list file that we MIGHT need to modify.
+    const pathToTestList = './cypress/TestLists.js';
+    var testList_jsContents = fs.readFileSync(pathToTestList, 'ascii');
+    // console.log(testList_jsContents)
+    // console.log()
+
+    // Get the list of test js files.
+    var testFileDirectory = fs.readdirSync(pathToTestFiles);
+    // console.log(testFileDirectory);
+    // console.log()
+
+    // Parse each test js file.
+    testFileDirectory.forEach(fileName => { ParseTestJsFile(fileName); });
+
+    // Finalize our findings by re-writing out the TestList.js file ONLY if it has changed.
+    var index = testList_jsContents.indexOf('// ************ Generated Code Beyond this Point *************************');
+    var newFileContents =
+      '// ************ Generated Code Beyond this Point *************************\r\n' +
+      '// Do NOT manually alter this file from this point onwards.\r\n' +
+      '// Any changes you make will be overridden at runtime.\r\n' +
+      'export const masterListOfAllTestCases = [\r\n';
+    fullTestList.forEach(testSpecification => { newFileContents += `  '${testSpecification}',\r\n` });
+    newFileContents += '];\r\n';
+
+    // Only write the file out if something has changed.
+    if (!testList_jsContents.endsWith(newFileContents)) {
+      newFileContents = testList_jsContents.substring(0, index) + newFileContents;
+      console.log(`TestList.js has been re-generated`)
+      fs.writeFileSync(pathToTestList, newFileContents)
+    }
+    else console.log('NOTHING HAS CHANGED')
+  }
+  catch(error) { console.log(`Caught an Exception:\r\n${error}`)}
+}
+
+// ----------------------------------------------------------------------
+
+function ParseTestJsFile(fileName) {
+  var fileContents = fs.readFileSync(`${pathToTestFiles}/${fileName}`, 'ascii')
+  //console.log(`File ${fileName}:\r\n${fileContents}`)
+  
+  var fileLines = fileContents.split('\n')
+  
+  fileLines.forEach(line => {
+    line = line.trim()
+    if (line.startsWith('Cypress.TestCase(')) {
+      var iStart = line.indexOf("'", 17) + 1
```

Comment:
Would at least factor this out into a variable so we can get remove at least some of coupling and hard-coding.

```
const attributeName = `Cypress.TestCase(`
if (line.startsWith(attributeName)) {
      var iStart = line.indexOf("'", attributeName.length) + 1
...
```

In my opinion a better solution for this type of searching would be to combine use of RegEx so people can see what parts of the line your extracting by looking at the expression.

Doesn't account all the possibilities, which again is why this solution is brittle, but it's an example:

```
const attributeRegex = /Cypress.TestCase\('(\w+)',\s'([\s\w]+)'/
...
const matches = line.match(attributeRegex)
if (matches) {
   const groupName = matches[1]
   const testName = matches[2]
   ...
}
```

Reply:
I don't like to use Regex due to its complexity and difficulty to read, unless not using it results in significantly slower performance or the resulting code is significantly more complex than this code is. And I do have a lot of experience using Regex, its just that you use it once, then you don't use it for 2 or 3 months again. Difficult to maintain an understanding of it when used so rarely.